applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_cut_pfem_process.hpp

```diff
@@ -0,0 +1,1706 @@
+
+			if (freesurfaceNodes > 0)
+			{
+				penalization = 1.2; // to avoid to gain too much volume during remeshing step
+			}
+			else if (negativeDistanceNodes > 0) // to penalize insertion of nodes near the boundary 
+			{
+				penalization = 0.9;
+			}
+
+			const double safetyCoefficient3D = 1.6;
+			array_1d<double, 6> Edges(6, 0.0);
+			array_1d<SizeType, 6> FirstEdgeNode(6, 0);
+			array_1d<SizeType, 6> SecondEdgeNode(6, 0);
+			double WallCharacteristicDistance = 0;
+			array_1d<double, 3> CoorDifference = Element[1].Coordinates() - Element[0].Coordinates();
+			double SquaredLength = CoorDifference[0] * CoorDifference[0] + CoorDifference[1] * CoorDifference[1] + CoorDifference[2] * CoorDifference[2];
+			Edges[0] = std::sqrt(SquaredLength);
+			FirstEdgeNode[0] = 0;
+			SecondEdgeNode[0] = 1;
+			if ((Element[0].Is(RIGID) && Element[1].Is(RIGID)) || (Element[0].Is(INLET) && Element[1].Is(INLET)))
+			{
+				WallCharacteristicDistance = Edges[0];
+			}
+			SizeType Counter = 0;
+			for (SizeType i = 2; i < nds; i++)
+			{
+				for (SizeType j = 0; j < i; j++)
+				{
+					noalias(CoorDifference) = Element[i].Coordinates() - Element[j].Coordinates();
+					SquaredLength = CoorDifference[0] * CoorDifference[0] + CoorDifference[1] * CoorDifference[1] + CoorDifference[2] * CoorDifference[2];
+					Counter += 1;
+					Edges[Counter] = std::sqrt(SquaredLength);
+					FirstEdgeNode[Counter] = j;
+					SecondEdgeNode[Counter] = i;
+					if (Element[i].Is(RIGID) && Element[j].Is(RIGID) && Edges[Counter] > WallCharacteristicDistance)
+					{
+						WallCharacteristicDistance = Edges[Counter];
+					}
+				}
+			}
+			// Edges connectivity: Edges[0]=d01, Edges[1]=d20, Edges[2]=d21, Edges[3]=d30, Edges[4]=d31, Edges[5]=d32
+			bool dangerousElement = false;
+			for (SizeType i = 0; i < 6; i++)
+			{
+				if (rigidNodes > 1)
+				{
+					if ((Edges[i] < WallCharacteristicDistance * safetyCoefficient3D && (Element[FirstEdgeNode[i]].Is(RIGID) || Element[SecondEdgeNode[i]].Is(RIGID))) ||
+						(Element[FirstEdgeNode[i]].Is(RIGID) && Element[SecondEdgeNode[i]].Is(RIGID)))
+					{
+						Edges[i] = 0;
+					}
+					if ((Element[FirstEdgeNode[i]].Is(FREE_SURFACE) || Element[FirstEdgeNode[i]].Is(RIGID)) &&
+						(Element[SecondEdgeNode[i]].Is(FREE_SURFACE) || Element[SecondEdgeNode[i]].Is(RIGID)))
+					{
+						Edges[i] = 0;
+					}
+				}
+				else if (rigidNodes == 0)
+				{
+					SizeType propertyIdFirstNode = Element[FirstEdgeNode[i]].FastGetSolutionStepValue(PROPERTY_ID);
+					SizeType propertyIdSecondNode = Element[SecondEdgeNode[i]].FastGetSolutionStepValue(PROPERTY_ID);
+					if (propertyIdFirstNode != propertyIdSecondNode)
+					{
+						penalization = 1.2; // 20% less than normal nodes
+					}
+				}
+			}
+			if (rigidNodes == 1)
+			{
+				if (Element[0].Is(RIGID))
+				{
+					Edges[0] = 0;
+					Edges[1] = 0;
+					Edges[3] = 0;
+				}
+				if (Element[1].Is(RIGID))
+				{
+					Edges[0] = 0;
+					Edges[2] = 0;
+					Edges[4] = 0;
```

Comment:
I see a lot of repetition, I suppose that you coudl reuse many methods but is up to you

Reply:
Agree. @AFranci I'll try to derive from the standard one to only reimplement what is strictly required.